### PR TITLE
lazily create an index on documents push

### PR DESCRIPTION
**Pull request by qdequele**
Create an index if it's possible when a user trying to send data to a non-existing index. https://github.com/meilisearch/MeiliSearch/issues/918

### CODE REVIEW

**Comment by Kerollmops** [comment]
Could you please add a link to a previous discussion about this feature? Like, do we all agreed, isn't there any cons about this feature?

**Comment by tpayet** [comment]
@qdequele was too fast in implementing this feature ^^

I personnaly think we should add this feature, but yes let's be sure we are not breaking anything ^^. Also, this feature needs to be documented 😉

File: meilisearch-http/src/routes/document.rs [hunk 1]
```diff
@@ -149,7 +181,8 @@ async fn update_multiple_documents(
     let index = data
         .db
         .open_index(&path.index_uid)
-        .ok_or(Error::index_not_found(&path.index_uid))?;
+        .ok_or(Error::index_not_found(&path.index_uid))
+        .or(create_index(&data, &path.index_uid))?;
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/document.rs:185]: This statement will be called always before every thing related to this line, you should use `or_else` instead.

**Review by Kerollmops — DISMISSED** [review]

File: meilisearch-http/src/data.rs [hunk 2]
```diff
@@ -95,4 +96,60 @@ impl Data {
...
+            None => {
+                index_has_been_created = true;
+                self.create_index(&uid)?
+            }
+        };
+
+        match f(&index) {
+            Ok(r) => Ok(r),
+            Err(err) => {
+                if index_has_been_created == true {
```
Comment (Kerollmops) [inline meilisearch-http/src/data.rs:148]:
```suggestion
                if index_has_been_created {
```

File: meilisearch-http/src/data.rs [hunk 3]
```diff
@@ -95,4 +96,60 @@ impl Data {
...
+                    let _ = self.db.delete_index(&uid);
```
Comment (Kerollmops) [inline meilisearch-http/src/data.rs:149]: You don't care about a potential error ?

Comment (qdequele) [reply meilisearch-http/src/data.rs:149]: yes, at the moment I don't care to know that the deletion of an index that I'm not sure that as been created failed.

**Comment by MarinPostma** [comment]
@qdequele can you resolve the conflicts please?

**Review by MarinPostma — DISMISSED** [review]
bors merge

**Comment by MarinPostma** [comment]
@bidoubiwa @curquiza @eskombro FYI

**Comment by LegendreM** [comment]
@qdequele can you update and push your `Cargo.lock` ?

**Comment by LegendreM** [comment]
bors try

**Review by LegendreM — APPROVED** [review]
bors r+

**Pull request merged**